### PR TITLE
Pattern Processor

### PR DESCRIPTION
**Pull request by stasm**
This is a continuation of #40. It adds a `Processor` class with two methods:

- `unescapeLiteralsToText` will bake string literals into surrounding `TextElements`,
- `escapeTextToLiterals` will do the reverse for characters which are special in Fluent syntax (e.g. `{` and `}`).

### CODE REVIEW

**Comment by Pike** [comment]
The idea about a separate data structure didn't work out? I'm not particularly opinionated, but if you can share what made you shift gears, that'd be interesting.

I've started looking at the patch, and docs and comments would help a lot, to tell what's in scope and what's not.

I saw in the tests that you process text inside select expressions, and I'm wondering if that's a good idea. Do we have a use-case for that?

**Comment by stasm** [comment]
> The idea about a separate data structure didn't work out? I'm not particularly opinionated, but if you can share what made you shift gears, that'd be interesting.

The main reason behind the decision to go with the Fluent AST was that it's already well-documented and has tools that allow working with it. Any new structure would mean new docs and likely new tools. It would also have a higher risk of becoming vendor-specific. 

I experimented with subclassing `Pattern` to make it explicit that it's a different kind of a pattern, one which may not be safely serialized without first un-processing it. However, to make this work, I'd need to either replicate the whole hierachy of `Pattern` → `Placeable` → `SelectExpression` → `Variant` → `Pattern`, or use casting. Casting seemed a bit arbitrary, and in fact not very different from just going with the actual Fluent AST, so I did that instead.

> I've started looking at the patch, and docs and comments would help a lot, to tell what's in scope and what's not.

Sorry for the PR being scarce on docs. I'll work on it.

> I saw in the tests that you process text inside select expressions, and I'm wondering if that's a good idea. Do we have a use-case for that?

I'd like the processing logic to be agnostic wrt. how variant permutations are handled. The expected way of handling them is through "exploding" all permutations into independent translation units, and this can happen after a pattern is processed. Does that answer your question?

**Comment by Pike** [comment]
> > I saw in the tests that you process text inside select expressions, and I'm wondering if that's a good idea. Do we have a use-case for that?
> 
> I'd like the processing logic to be agnostic wrt. how variant permutations are handled. The expected way of handling them is through "exploding" all permutations into independent translation units, and this can happen after a pattern is processed. Does that answer your question?

Maybe. It's another thing to think about during the review.

Also, in the case of a shall-not-explode string, we should make sure that we're exposing the untransformed message content, and not the transformed one. At least I think so, as the transformed one would be hard to round-trip, right?

**Comment by stasm** [comment]
> Also, in the case of a shall-not-explode string, we should make sure that we're exposing the untransformed message content, and not the transformed one. At least I think so, as the transformed one would be hard to round-trip, right?

Can you rephrase this, please? What is a _shall-not-explode_ string, and expose to whom? :)

**Comment by Pike** [comment]
There are messages that some l10n tools may choose to not explode, but show to localizers as plain syntax instead. They should show the original message source, and not a transformed message source. At least that'd be my expectation.

**Comment by stasm** [comment]
Hmm, interesting. If a message is `key = Hello, {"{-_-}"}.`, my understanding is that it doesn't require exploding, and that a text-oriented editor should display its value processed, as `Hello, {-_-}.` If I understand correctly, that's not what you'd expect?

**Comment by stasm** [comment]
I added docstrings in the latest commit. I hope they help with the review!

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                    }
+                }
+                is Placeable -> {
+                    when (val expression = element.expression) {
+                        is SelectExpression -> {
+                            val rawVariants: MutableList<Variant> = mutableListOf()
+                            for (variant in expression.variants) {
+                                val rawVariant = Variant(variant.key, escapeTextToLiterals(variant.value), variant.default)
+                                rawVariants.add(rawVariant)
+                            }
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:210]: Same here as for the variants above.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 2]
```diff
@@ -0,0 +1,238 @@
+package org.projectfluent.syntax.processor
+
+import org.projectfluent.syntax.ast.* // ktlint-disable no-wildcard-imports
+import java.lang.Exception
+
+private val special =
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:6]: Can we put this and `unescape` next to each other?

Also, this doesn't handle `\U123456` yet. I know it's my fault for not doing so initially, but good to call out.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    private fun unescape(matchResult: MatchResult): CharSequence {
+        val matches = matchResult.groupValues.drop(2).listIterator()
+        val simple = matches.next()
+        if (simple != "") { return simple }
+        val uni4 = matches.next()
+        if (uni4 != "") {
+            val a = uni4.substring(1).toInt(16)
+            val b = a.toChar()
+            val c = b.toString()
+            val d = a.toString()
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:233]: I've had these to step through in the debugger, these should be removed now.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                is TextElement -> {
+                    if (lastText == null) {
+                        lastText = element
+                    } else {
+                        lastText?.let { it.value += element.value }
+                    }
+                }
+                is Placeable -> {
+                    when (val expression = element.expression) {
+                        is NumberLiteral -> {
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:109]: The `NumberLiteral` handling doesn't have tests and docs yet?

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                                lastText = TextElement("")
+                            }
+                            lastText?.let { it.value += content }
+                        }
+                        is SelectExpression -> {
+                            val processedVariants: MutableList<Variant> = mutableListOf()
+                            for (variant in expression.variants) {
+                                val processedVariant = Variant(variant.key, unescapeLiteralsToText(variant.value), variant.default)
+                                processedVariants.add(processedVariant)
+                            }
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:129]: I bet the variants can be written as a sequence and `.toMutableList()`, but I don't know how exactly.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+     *      {"*"} Asterisk is special
+     *     ←A multiline pattern:
+     *      * Asterisk is special
+     *
+     *     →Copyright {"\u00A9"} 2020
+     *     ←Copyright © 2020
+     */
+    fun unescapeLiteralsToText(pattern: Pattern): Pattern {
+        val result = Pattern()
+        for (elem in textFromLiterals(pattern)) {
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:60]: Both here and below, you can probably use `result.elements.addAll(textFromLiterals(pattern))`?

**Review by Pike — CHANGES_REQUESTED** [review]
Nice docs, I think the majority of it should be on a page in the docs folder, though. I'd like to see the docstrings focus on just a tldr. I locally looked at the tooltips in IDEA on the test, and that's a screenful ;-).

Some review comments I left inline are my fault, sorry for that. It'd still be good to clean those up.
I also recognized that I left quite a few local variables that are useful when debugging, but don't help the code in readability. Not a  big one, though.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 2 above)
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:6]: I moved `val special` into the class body, and I filed #44 for the `\U123456` case.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 5 above)
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:129]: Let's keep this as-is, then. Boring code is good :)

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            when (element) {
+                is TextElement -> {
+                    if (element.value.startsWith(' ') || element.value.startsWith('\n')) {
+                        val expr = StringLiteral("")
+                        yield(Placeable(expr))
+                    }
+
+                    var startIndex = 0
+                    var endIndex = 0
+                    for (i in element.value.indices) {
```
[the remark this answers is not in the document]
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:164]: I think so, but I think it wouldn't work with `value.indices` then?

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 6 above)
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:16]: I think I prefer the good old `for` loop, unless there are functional differences?

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        """\\(([\\"])|(u[0-9a-fA-F]{4}))""".toRegex()
+
+    private fun unescape(matchResult: MatchResult): CharSequence {
+        val matches = matchResult.groupValues.drop(2).listIterator()
+        val simple = matches.next()
+        if (simple != "") { return simple }
+        val uni4 = matches.next()
+        if (uni4 != "") {
+            val a = uni4.substring(1).toInt(16)
+            val b = a.toChar()
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:172]: `a` and `b` are also unused.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pike) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:102]: On a second thought, you don't need  `endIndex` at all, as it's always `element.value.lastIndex + 1`

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                                endIndex = i + 1
+                            }
+                            else -> {
+                                endIndex = i + 1
+                            }
+                        }
+                    }
+
+                    // Yield the remaining text.
+                    if (endIndex > startIndex) {
```
Comment (Pike) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:131]: `element.value.lastIndex >= startIndex` here, and then just `substring(startIndex)` below.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 4 above)
Comment (Pike) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:47]: I'm leaning towards actually keeping  `NumberLiteral`s in the processed pattern, and mark it up as protected placeable.

Rationale  being that `{20000}`  renders differently in different locales, and is different to `20000` in most languages?

**Review by Pike — CHANGES_REQUESTED** [review]
Thanks for the docs, I built and reviewed them locally.

There's a few small things still, and I've changed my mind on the `NumberLiteral` thing. Which is good, as it means less docs and tests to do ;-) I think we should just let them go through the processor unchanged.

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 8 above)
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:172]: D'oh, of course they are, thanks :)

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:102]: Oh, you're right, nice!

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt (hunk 4 above)
Comment (stasm) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/processor/Processor.kt:47]: This is a good insight, thanks. I agree, let's leave them in.

**Review by Pike — APPROVED** [review]
Sweet, thanks for taking this. r=me.
